### PR TITLE
Make explicit the tuple type for strategies

### PR DESCRIPTION
**Pull request by iwahbe**
Even though this is technically a breaking change, I only expect the `datadog` provider to require a fix.

### CODE REVIEW

File: pkg/tfbridge/info.go [hunk 1]
```diff
@@ -99,39 +99,47 @@ type ProviderInfo struct {
...
 	}
 	return errs.ErrorOrNil()
 }
 
-// Apply strategy to generate missing resources.
-//
-// NOTE: Experimental; We are still iterating on the design of this function, and it is
-// subject to change without warning.
-func (info *ProviderInfo) ComputeDefaultResources(strategy ResourceStrategy) error {
+func (info *ProviderInfo) computeDefaultResources(strategy ResourceStrategy) error {
```
Comment (iwahbe) [inline pkg/tfbridge/info.go:129]: Now that we have a struct for passing strategies, we never have a reason to expose these helpers. We may as well remove them during this phase.

File: pkg/tfbridge/token.go [hunk 2]
```diff
@@ -34,13 +34,13 @@ type ResourceStrategy = Strategy[ResourceInfo]
...
 // subject to change without warning.
-type DatasourceStrategy = Strategy[DataSourceInfo]
+type DataSourceStrategy = Strategy[DataSourceInfo]
 
 // A generic remapping strategy.
 //
 // NOTE: Experimental; We are still iterating on the design of this type, and it is
 // subject to change without warning.
-type Strategy[T ResourceInfo | DataSourceInfo] func(tfToken string, tfTokens []string) (*T, error)
+type Strategy[T ResourceInfo | DataSourceInfo] func(tfToken string) (*T, error)
```
Comment (iwahbe) [inline pkg/tfbridge/token.go:43]: No strategy ever used the "all tokens" argument, so I'm remove it. Again, while this is technically breaking, no providers will be impacted.

File: pkg/tfbridge/token.go [hunk 3]
```diff
@@ -121,40 +121,55 @@ func tokensKnownModules[T ResourceInfo | DataSourceInfo](
...
+				tk, err := finalize(mod, "get"+tk)
+				if err != nil {
+					return nil, err
+				}
+				return &DataSourceInfo{Tok: tokens.ModuleMember(tk)}, nil
+			}),
+	}
+}
+
+func (ts ComputeDefaultInfo) Unmappable(substring, reason string) ComputeDefaultInfo {
```
Comment (iwahbe) [inline pkg/tfbridge/token.go:150]: In practice, marking something as unmappable without explaining why leads to confusion later down the line. I'm augmenting this API to allow the user to explain why something is unmappable.

File: pkg/tfbridge/info.go (hunk 1 above)
Comment (t0yv0) [reply pkg/tfbridge/info.go:129]: Yeah! I think either use a struct or remove ComputeDefaults completely and only provide the individual helpers, keep surface area a bit smaller! I think the struct is appropriate as there are the multierror helpers baked in.

File: pkg/tfbridge/token.go (hunk 3 above)
Comment (t0yv0) [reply pkg/tfbridge/token.go:150]: SGTM.

**Review by t0yv0 — APPROVED** [review]

File: pkg/tfbridge/info.go [hunk 4]
```diff
@@ -99,39 +99,47 @@ type ProviderInfo struct {
 	PreConfigureCallbackWithLogger PreConfigureCallbackWithLogger
 }
 
+// Describe the mapping from resource and datasource tokens to Pulumi resources and
+// datasources.
+//
+// NOTE: Experimental; We are still iterating on the design of this function, and it is
+// subject to change without warning.
+type ComputeDefaultInfo struct {
```
Comment (t0yv0) [inline pkg/tfbridge/info.go:107]: Nit: I don't like the *Info name. It puts it as a synonym of ResourceInfo in my head which is not as helpful. Perhaps. DefaultsStrategy?

Comment (iwahbe) [reply pkg/tfbridge/info.go]: Changed in https://github.com/pulumi/pulumi-terraform-bridge/pull/766/commits/8145571811581e9e848bcaccbba2527b956079a0

**Pull request merged**